Allow plugin to be disabled

bpinto opened the pull request:
This allows jest to run tests since last branch or since last commit by adding a `enabled` flag to the plugin.

>  › Press m to test changes since master.
>  › Press m to test changes since last commit.

bpinto committed:
Allow plugin to be disabled

rickhanlonii commented:
@bpinto thanks! It's not clear to me how this gets enabled/disabled, can you explain how this works?

bpinto commented:
Jest watch will by default run all tests related to code changes since last commit, and `jest-watch-master` adds the option to run all tests related to code changes since master.

The toggle I've added allows one to go back and forth between the default mode and this plugin mode, in other words, to enable or disable this plugin.

For someone that continuously commit - using fixup most often - not being able to disable the plugin can be disrupting, since jest watch needs to be restarted in order to disable this plugin.

From jest docs:

````
jest --watch #runs jest -o by default
````

> Run tests related to changed files based on hg/git (uncommitted files):

````
jest -o
````

rickhanlonii commented:
But this plugin doesn't change the default behavior, it only adds a different watch mode option:

```
Watch Usage
 › Press a to run all tests.
 › Press f to run only failed tests.
 › Press o to only run tests related to changed files.
 › Press m to test changes since master.
 › Press p to filter by a filename regex pattern.
 › Press t to filter by a test name regex pattern.
 › Press q to quit watch mode.
 › Press Enter to trigger a test run.
```

So there are two options based on what you want to do:

```
 › Press o to only run tests related to changed files.
 › Press m to test changes since master.
```

Am I missing something?

bpinto commented:
🤔 This is what I have:

```
Watch Usage
 › Press a to run all tests.
 › Press f to run only failed tests.
 › Press p to filter by a filename regex pattern.
 › Press t to filter by a test name regex pattern.
 › Press q to quit watch mode.
 › Press m to test changes since master.
 › Press Enter to trigger a test run.
```

That explains why you don't need it but I do, I'm investigating what's going on.

bpinto commented:
Ah, found the reason:

https://github.com/facebook/jest/blob/9ffd368330a3aa05a7db9836be44891419b0b97d/packages/jest-core/src/watch.ts#L546-L553

Do you happen to run jest with the `--watchAll` flag? I start jest with `--watch`.

```
(globalConfig.watchAll || globalConfig.testPathPattern || globalConfig.testNamePattern)
```

This is the check that `jest` checks before displaying or not the `o` option, it doesn't take into consideration if this plugin is enabled or not.

Perhaps instead of changing what this plugin does, I could submit a PR to add an additional check that `jest` would consider for displaying the `o` option (if `changed-since` is set).

rickhanlonii commented:
@bpinto no, I just use `--watch`, are you running the tests with a particular file selected?

Yeah I think if there's a bug affecting your workflow here then the bug is in Jest and not this plugin.

Going to close, but keep me updated!
